### DISCUSSION TITLE
Freezing URL-based libraries during firmware compilation

### DISCUSSION THREAD

**Discussion by fill1890** [micropython/micropython#15759] [Libraries & Drivers] [answered]
`mip` can install libraries directly from GitHub (or other online sources) by passing a URL and using a `package.json` file. Is it possible to do the same thing in a `manifest.py` file? I know I can use `require()` to install directly from `micropython-lib`, but I have other libraries that are hosted remotely I'd like to freeze into the firmware.

Currently I clone the GitHub repos and use `package()` or `module()`, but this becomes tedious when I need to install multiple libraries.

Thanks!

**Comment by bidrohini** [comment]
Unfortunately, there is no built-in feature in MicroPython that allows you to directly install libraries from a URL or GitHub repository within the manifest.py file, as you can with package.json in Node.js.

**Comment by Josverl** [comment, marked as answer]
If you are already use `git clone` to get those resources in , then I suggest you take a look at [`git submodule`](https://git-scm.com/docs/git-submodule) to automate this for you.
You will need to get used to a few additional commands, and 'pin' specific versions, commits or branches of your dependencies  , but once you do, it can work quite well.

This is a quite common workflow , and is also used by MicroPython itself to pull in the dependencies needed for a build by running the `make submodules` command as part of the build.

See : 
- https://github.com/micropython/micropython/blob/master/.gitmodules
- the `make submodules` command and the `submodules:` section in the different makefiles.
